include/fiction/networks/virtual_pi_network.hpp

```diff
@@ -0,0 +1,693 @@
+                        continue;
+                    }
+                }
+            }
+        }
+
+        ntk.foreach_po(
+            [&ntk, &ntk_dest_v, &old2new_v](const auto& po)
+            {
+                const auto tgt_signal_v = old2new_v[ntk.get_node(po)];
+                // POs do not get duplicated since the algorithm starts at the POs and duplicates other nodes according
+                // to their order
+                assert(tgt_signal_v.size() == 1);
+                const auto tgt_signal = tgt_signal_v[0];
+
+                const auto tgt_po = ntk.is_complemented(po) ? ntk_dest_v.create_not(tgt_signal) : tgt_signal;
+
+                ntk_dest_v.create_po(tgt_po);
+            });
+
+        *this = ntk_dest_v;
+    }
+
+    /**
+     * @brief Reimplementation of num_gates(). Retrieves the number of gates in the virtual_pi_network.
+     *
+     * @return The number of gates in the virtual_pi_network.
+     */
+    auto num_gates() const
+    {
+        return static_cast<uint32_t>(_storage->nodes.size() - virtual_inputs->size() - _storage->inputs.size() - 2);
+    }
+
+    /**
+     * @brief Reimplementation of size(). Get the size of the virtual_pi_network.
+     *
+     * @return The size of the virtual_pi_network as a uint32_t.
+     */
+    auto size() const
+    {
+        return static_cast<uint32_t>(_storage->nodes.size());
+    }
+
+    /**
+     * \brief Calculate the real size of the virtual_pi_network.
+     *
+     * This function calculates the real size of the virtual_pi_network by subtracting the size of the virtual_inputs
+     * vector from the size of the _storage->nodes vector.
+     *
+     * \return The real size of the virtual_pi_network as a uint32_t.
+     */
+    [[nodiscard]] auto size_real() const
+    {
+        return static_cast<uint32_t>(_storage->nodes.size() - virtual_inputs->size());
+    }
+
+    /**
+     * @brief Create a virtual PI mapping to a real PI.
+     *
+     * This function creates a virtual PI mapping to a real PI in the network. It adds a new node to the
+     * `_storage->nodes` vector and inserts the index of the new node into the `virtual _inputs` vector. It also updates
+     * the data of the new node.
+     *
+     * @param real_pi The node representing the real PI in the network.
+     * @return The index of the newly created virtual PI.
+     */
+    signal create_virtual_pi(const node& real_pi)
+    {
+        const auto index = _storage->nodes.size();
+        _storage->nodes.emplace_back();
+        virtual_inputs->emplace_back(index);
+        _storage->nodes[index].data[1].h1 = 2;
+        map.emplace_back(index, real_pi);
+        return index;
+    }
+
+    /**
+     * @brief Reimplementation of is_pi(). Check if a given node is a primary input (PI) in the virtual_pi_network.
+     *
+     * A node is considered a PI if it exists in the virtual_inputs vector or the inputs vector of the _storage object.
```

Comment:
This is another good example of *how* vs. *what*. I would reimagine this docstring as follows:
```suggestion
     * A node is considered a PI if it is a virtual PI created via `create_virtual_pi()` or a real PI created in the original network via `create_pi()`.
```